### PR TITLE
[RFC] Issue #9104 Provide dialog widgets

### PR DESCRIPTION
**Pull request by yybalam**
This PR try to provide to qooxdoo a default basic dialog widgets. Basically this PR add :
1. A Message dialog for alert messaged.
2. A Confirm dialog with support for attach function to buttons.
3. A BusyIndicator.

A `qx.ui.Dialog` static class provide a short way for launch dialogs.:

```
qx.ui.Dialog.alert("Yay, some dialog!");
qx.ui.Dialog.error("Some was wrong. :(");
```

Also this PR include an `MBusyBlocker` mixin. This mixin extend `MBlocker` for provide an loading gif while the widget is blocked. Also is possible to show a custom text if `allowBusyPopup` is set to `true`.

Right now there is only two test, one for `Message` and another for `Confirm`.
## Whats missing?

Well, some documentation of course. Also I thing a test for  `MBusyBlocker`, but I don found a test for `MBlocker` so I'm not sure how to test it.

Also when I compile using 'Tango' icon theme, the icons are shown, but I can't get it work for Oxygen.

I know this PR actually is not complete (sorry) to be merge, but maybe heps to get the way.

Cheers.

### CODE REVIEW

**Comment by cajus** [comment]
It is better to use the qx childControl mechanism for widgets, buttons, etc. and move everything styling related to the appearances/decorators. Additionally the dialogs expose the complete children handling to the outside world (add/remove/...). Not sure if we want this.

Besides the fact that you don't have to manually dispose the widgets you've created (missing in your code), you've the advantage that it is completely styleable by dialog users if you use child-controls. No need for custom methods like _getButtonBar, _getAtom, etc.

As dialogs are "one shot", so it's maybe a good idea to add an *autodispose* flag to the dialog base class.

File: framework/source/class/qx/ui/dialog/Abstract.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    this.add(this._getAtom());
+    this.add(this._getButtonsBar());
+
+    this._initDialog(title, message);
+
+    if (icon != undefined) {
+      this._getAtom().setIcon(icon);
+    }
+  },
```
Comment (oetiker) [inline framework/source/class/qx/ui/dialog/Abstract.js]: you may want to follow the structure outlined [in the qooxdoo manuel](http://www.qooxdoo.org/current/pages/desktop/ui_develop.html) using ChildControls

File: framework/source/class/qx/ui/dialog/Abstract.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+     */
+    _initDialog : function(title, message) {
+      this.__title = title;
+      this.setMessage(message);
+    },
+
+    /**
+     * @param title {String} Title of dialog.
+     */
+    setTitle : function(title) {
```
Comment (johnspackman) [inline framework/source/class/qx/ui/dialog/Abstract.js]: `title` and `message` would be better implemented as a property

**Comment by johnspackman** [comment]
looking good @yybalam

**Comment by oetiker** [comment]
you may want to have a look at the color popup for inspiration https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/ui/control/ColorPopup.js

**Comment by oetiker** [comment]
looks much more standard now ... you had some live demo, where is it ?

also for integration we need updates to the manual and ideally also to the demobrowser ... 

and you should look into squashing the many commits

**Comment by yybalam** [comment]
Hello.

Here is a live demo: http://yybalam.net/qxlab/dialoglab/
Here is the code of the demo: https://gitlab.com/yybalam/dialoglab/tree/master

So, I will to write the docs for manual and examples to demobrower, also add support for the rest of themes and the squashing of my lot of commits (at moment I don't kno how to do this last one, need to study :sweat_smile:)

**Comment by woprandi** [comment]
@yybalam you can check how to squash commits here : https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Squashing-Commits

**Comment by oetiker** [comment]
looking at the demo ... like the optics ... I wonder about the placement of the buttons ... I would have expected them to be right aligned.

File: framework/source/class/qx/bom/Cookie.js [hunk 3]
```diff
@@ -62,7 +62,7 @@ qx.Bootstrap.define("qx.bom.Cookie",
         end = document.cookie.length;
       }
 
-      return decodeURI(document.cookie.substring(len, end));
+      return unescape(document.cookie.substring(len, end));
     },
```
Comment (oetiker) [inline framework/source/class/qx/bom/Cookie.js]: why ?

File: framework/source/class/qx/bom/Cookie.js [hunk 4]
```diff
@@ -80,7 +80,7 @@ qx.Bootstrap.define("qx.bom.Cookie",
     set : function(key, value, expires, path, domain, secure)
     {
       // Generate cookie
-      var cookie = [ key, "=", encodeURI(value) ];
+      var cookie = [ key, "=", escape(value) ];
```
Comment (oetiker) [inline framework/source/class/qx/bom/Cookie.js]: why ?

File: framework/source/class/qx/bom/Cookie.js (hunk 3 above)
Comment (yybalam) [inline framework/source/class/qx/bom/Cookie.js]: Nope. That is fixed by another PR that is no merged yet, so I think this PR must avoid that change, so this PR don't break anything with actual sate of master. :sweat_smile:

Comment (oetiker) [inline framework/source/class/qx/bom/Cookie.js]: ah this is a revert to your own patch series ... you definitely need to squash these bevor we can merge ...

Comment (yybalam) [inline framework/source/class/qx/bom/Cookie.js]: Yes. I'll try to do it right in my next try. Squash is absolutely something new for me. :sweat_smile:

**Comment by zaucker** [comment]
Are the block and busy variants styled differently than the other dialogs on purpose in the demo? Title and min/max/close buttons on block/busy (extending qx.ui.window.Window() and not on the others (probably extending Popup)?

**Comment by yybalam** [comment]
Yeah. The idea behind bolck and busy is a mixin (MBusyBlocker) for any widget, so do not have any special style, is just for block a Widget and shown a centred gif or popup with a tiny gif and a label. I just used Windows for convenience in the demo.

**Comment by zaucker** [comment]
Thanks for the explanation, sounds good.

On Wed, 31 Aug 2016, Yeye Balam wrote:

> Yeah.  The idea behind bolck and busy is a mixin (MBusyBlocker) for any
> widget, so do not have any special style, is just for block a Widget and
> shown a centred gif or popup with a tiny gif and a label.  I just used
> Windows for convenience in the demo.

-- 
Oetiker+Partner AG              tel: +41 62 775 9903 (direct)
Fritz Zaucker                        +41 62 775 9900 (switch board)
Aarweg 15                            +41 79 675 0630 (mobile)
CH-4600 Olten                   fax: +41 62 775 9905
Schweiz                         web: www.oetiker.ch

**Comment by cboulanger** [comment]
Hello @yybalam, still interested in contributing a widget set to qooxdoo for the 6.0 release? This would require that you update the PR to the current master and address the remaining issues raised in the discussion. You could also contribute your widget set as a "contrib", which would require to change the top-level namespace. It would be great to have two different dialog widget sets as contribs, so people have a choice... If you're interested, let me know.

**Comment by yybalam** [comment]
@cboulanger Oh, yeah. I'm still interesed in contributing the widget. But instead of update this PR, can I close this PR and create a new one?

**Comment by cboulanger** [comment]
@yybalam Sure!

**Comment by level420** [comment]
As @yybalam suggested: closing this PR.

**Comment by cboulanger** [comment]
@yybalam please consider creating a contrib based on a child controls. It would be great to have stylable dialogs, I don't think I will have the time to ever work on my own Dialog widget and add styleability. Publishing the contrib will be really easy with the new `qx contrib` CLI commands which I hope to finish soon.

**Pull request closed without merging**